### DISCUSSION TITLE
Is MVU still planned for .net MAUI?

### DISCUSSION THREAD

**Discussion by ske66** [dotnet/maui#1472] [General]
I remember seeing the initial demos of MAUI in 2020 showing off the ability to use the MVU architecture for your projects. Since those early demos I haven't really seen any material regarding MVU on MAUI, neither in articles, demos, or any of the previews. Is this still a planned feature? I saw some discussion regarding MVU in Issue 118 where people were saying that the implementation of MVU shown in the demos was not a realistic representation of MVU in c#, more akin to psuedo code.

So are there still plans for MVU to be available in MAUI? I develop a lot of my web apps in React, and sometimes develop mobile apps with React Native, so would be good to see MVU used for the Mobile side of development in MAUI at least.

**Comment by saint4eva** [comment]
> I remember seeing the initial demos of MAUI in 2020 showing off the ability to use the MVU architecture for your projects. Since those early demos I haven't really seen any material regarding MVU on MAUI, neither in articles, demos, or any of the previews. Is this still a planned feature? I saw some discussion regarding MVU in Issue 118 where people were saying that the implementation of MVU shown in the demos was not a realistic representation of MVU in c#, more akin to psuedo code.
> 
> So are there still plans for MVU to be available in MAUI? I develop a lot of my web apps in React, and sometimes develop mobile apps with React Native, so would be good to see MVU used for the Mobile side of development in MAUI at least.

The .NET Maui MVU implementation is going to be Comet for C# and Fabolous for F#. For Comet - which was demoed last year - the repo is hosted here https://github.com/dotnet/Comet

**Reply by ske66** [reply]
So they are going to be kept as separate entities rather than shipped with MAUI on release?

**Reply by saint4eva** [reply]
It is being worked on. The apis are changing and need to be stable before it can be brought into .NET Maui - mind you Maui is still being developed and things are changing also. James Clancey is expecting the C# team to add some features to C# 10 that would make Comet exceptionally excellent to write UI with.

**Reply by ske66** [reply]
that sounds great, I love working with MVU so would be awesome to start developing mobile apps with it Just such a better developer experience all around

**Comment by dgellow** [comment]
There is now a library for MAUI MVU, `CommunityToolkit.Maui.Markup`: https://github.com/communitytoolkit/Maui.Markup.

**Reply by saint4eva** [reply]
Maui Markup is a fantastic library for writing beautiful UIs in C#. However, it leans towards MVVM than MVU. For MVU C#, you should try:
https://github.com/dotnet/Comet

**Reply by dgellow** [reply]
You're correct, I found out about Comet after posting my previous message :)
From what I understand, after a short chat with Comet's author, differences are the following:
- CommunityToolkit.Maui.Markup offers a set of extension methods to make it convenient to create Maui controls (from `Microsoft.Maui.Controls`) directly from C#, without any XAML. It's a convenience wrapper for Maui in some way.
- Comet brings a state management system, a hot reload system, and implement the Maui control handler interfaces (from `Microsoft.Maui`). So, while it does replace the Markup library, that's more a side effect. It's more an alternative to `Microsoft.Maui.Controls`. 

Using the diagram from [Maui documentation](https://docs.microsoft.com/en-us/dotnet/maui/user-interface/handlers/customize), that can be described like the following:

![image](https://user-images.githubusercontent.com/2451004/142878127-5c8b0112-4708-4069-8981-08dae0007edb.png)

I haven't tried Maui.Markup yet, and do not have an opinion yet. But I can say that Comet seems to offers exactly the type of developer experience I would expect to have in 2021. Kudos to @Clancey and other contributors.

**Reply by saint4eva** [reply]
Yes. Comet is an app model untop of Maui. So, Comet, Fabulous, Control and Blazor are all appmodel untop of dotnet Maui.
